File: python/leetcode/2_add_two_numbers.py
"""
2. Add Two Numbers

You are given two non-empty linked lists representing two non-negative integers.
 The digits are stored in reverse order and each of their nodes contain a single digit.
  Add the two numbers and return it as a linked list.

You may assume the two numbers do not contain any leading zero, except the number 0 itself.

Example:

Input: (2 -> 4 -> 3) + (5 -> 6 -> 4)
Output: 7 -> 0 -> 8
Explanation: 342 + 465 = 807.
"""

"""
Time : (more than) 60 min

* Subject - Linked List
* Learning
    1. Should start from the 'last node' when making linked list in python
    2. The place of 'break' when make a for loop for linked list
        181201
        -> Just using 'while node:' instead
    
    3. Power of pseudo code, simple
    
"""

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class ListNode:
    def __init__(self, val):
        self.val = val
        self.next = None


class Solution:
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        # l1_num = 0
        # l2_num = 0
        #
        # for idx, i in enumerate(l1):
        #     # if idx == 0:
        #     #     l1_num += i
        #     # else:
        #     #     l1_num = l1_num + i * 10 ** idx
        #     l1_num = l1_num + i * 10 ** idx if not idx == 0 else l1_num + i
        #
        # for idx, i in enumerate(l2):
        #     l2_num = l2_num + i * 10 ** idx if not idx == 0 else l2_num + i
        #
        # sum = l1_num + l2_num
        #
        # sum_list = []
        # for i in str(sum):
        #     sum_list.insert(0, i)
        #
        # return sum_list

        def print_node(l):
            while True:
                print(l.val, end=' -> ') if l.next else print(l.val)
                if l.next is None:
                    break
                l = l.next

        l1_num = 0
        l2_num = 0

        i = 0
        while True:
            l1_num = l1_num + l1.val * 10 ** i if not i == 0 else l1_num + l1.val
            i += 1
            if not l1.next:
                break
            l1 = l1.next

        i = 0
        while True:
            l2_num = l2_num + l2.val * 10 ** i if not i == 0 else l2_num + l2.val
            i += 1
            if not l2.next:
                break
            l2 = l2.next

        sum = l1_num + l2_num
        sum_str = str(sum)

        node = ListNode(sum_str[0])
        for idx, i in enumerate(sum_str[1:]):
            next_node = ListNode(int(i))
            next_node.next = node
            node = next_node

        print_node(node)
        return node

        # * Fail - cannot find root node after the process

        # reverse_sum = []
        # for i in str(sum):
        #     reverse_sum.insert(0, int(i))
        #
        # l3 = ListNode(reverse_sum[0])
        #
        # for i in reverse_sum[1:]:
        #     node = ListNode(i)
        #     l3.next = node
        #     l3 = l3.next


if __name__ == '__main__':

    l1 = ListNode(2)
    l2 = ListNode(4)
    l3 = ListNode(3)
    l1.next = l2
    l2.next = l3

    l4 = ListNode(5)
    l5 = ListNode(6)
    l6 = ListNode(4)
    l4.next = l5
    l5.next = l6

    solution = Solution()
    solution.addTwoNumbers(l1, l4)


"""

181201 Review

Time : 13 min

The code below is simpler than the above in two ways.
First, the one above make a non reverse number by multiplying 10**i.
       it is not easy to think instantly.
Second, no need to make a last node outside of loop like 'node = ListNode(sum_str[0])'
      There is better way as the code below using 'prev_node=None'

"""

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """

        def make_num(node):
            str_num = ''
            while node:
                str_num = str(node.val) + str_num
                node = node.next

            return int(str_num)

        num1 = make_num(l1)
        num2 = make_num(l2)

        num3_str = str(num1 + num2)

        node = None
        prev_node = None
        for i in num3_str:
            node = ListNode(int(i))

            if prev_node:
                node.next = prev_node
            prev_node = node

        return node


def make_node(num):

    str_num = str(num)

    node = None
    prev_node = None
    for i in str_num:
        node = ListNode(int(i))

        if prev_node:
            node.next = prev_node
        prev_node = node

    return node


num1 = 342
num2 = 465
l1 = make_node(num1)
l2 = make_node(num2)


solution = Solution()
l3 = solution.addTwoNumbers(l1, l2)

node = l3
while node:
    print(node.val, end=' -> ' if node.next is not None else '')
    node = node.next

File: javascript/leetcode/2_add_two_numbers.js
/**
 * Definition for singly-linked list.
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */

function ListNode(val) {
    this.val = val;
    this.next = null;
}


// 1) (fail)

function makeNum(llist) {
    let num = 0;
    while (true) {
        let i = 0
        if (llist) {
            num += llist.val * (10 ** i);
            llist = llist.next;
            ++i;
        }
        else {
            break;
        }
    }
    return num;
}
function makeLlist(value) {
    let prevNode = null;
    let node = null;
    for (var i of value.toString()) {
        node = new ListNode(i)
        if (prevNode) {
            node.next = prevNode;
        }
    }
    return node;
}
var addTwoNumbers = function(l1, l2) {
    console.log(1, makeNum(l1))
    console.log(2, makeNum(l2))
    const sumNum = makeNum(l1) + makeNum(l2)
    console.log(3, sumNum)
    return makeLlist(sumNum)
};



// 200331
// 2) (fail)
// result = addTwoNumbers(makeLinkedList(1000000000000000000000000000001), makeLinkedList(465))

function ListNode(val) {
    this.val = val;
    this.next = null;
}

function printLinkedList(l) {
    let numStr = ''
    while(true) {
        numStr += l.val
        if (!l.next) break;
        numStr += ' -> '
        l = l.next
    }
    console.log(numStr)
}

function makeLinkedList(num) {
    numList = [];
    idx = 1;
    while(true) {
        numList.unshift(num % 10);
        num = Math.floor(num / 10);
        if (num < 1) break;
        idx++;
    }
    console.log(2, numList)

    // const numStr = String(num);
    let prev;
    for(let i=0; i < numList.length; i++) {
        l = new ListNode(numList[i])
        if (prev) l.next = prev
        prev = l
    }
    return l
}

function makeNum(l) {
    let num = 0;
    let idx = 1;
    while(true) {
        num += l.val * 10 ** (idx - 1);
        if (!l.next) break;
        l = l.next;
        idx++;
    }
    return num;
}

var addTwoNumbers = function(l1, l2) {
    num3 = makeNum(l1) + makeNum(l2)
    console.log(num3)
    l3 = makeLinkedList(num3)
    return l3
};

printLinkedList(makeLinkedList(342))
printLinkedList(makeLinkedList(465))
result = addTwoNumbers(makeLinkedList(342), makeLinkedList(465))
printLinkedList(result)

// printLinkedList(makeLinkedList(10000000000000000001))
// printLinkedList(makeLinkedList(465))
// result = addTwoNumbers(makeLinkedList(1000000000000000000000000000001), makeLinkedList(465))
// printLinkedList(result)




// 200401
// 3) (retrial)

function ListNode(val) {
    this.val = val;
    this.next = null;
}

function printLinkedList(l) {
    let numStr = ''
    while(true) {
        numStr += l.val
        if (!l.next) break;
        numStr += ' -> '
        l = l.next
    }
    console.log('(', numStr, ')')
}

function makeLinkedList(value) {
    let prevNode = null;
    let node = null;
    for (var i of value.toString()) {
        node = new ListNode(Number(i))
        if (prevNode) {
            node.next = prevNode;
        }
        prevNode = node;
    }
    return node;
}

// ----------------------------------------------------

function makeLinkedListFromArr(arr) {
    let prev;
    for(let i=arr.length-1; i >= 0; i--) {
        node = new ListNode(arr[i]);
        if (prev) node.next = prev;
        prev = node;
    }
    return node;
}

function arraySum(arr1, arr2) {
    let sumArr = [];
    [arrMain, arrSub] = arr1.length > arr2.length ? [arr1, arr2] : [arr2, arr1]

    for (let i = 0; i < arrMain.length; i++) {
        // arrMain + arrSub
        if (i < arrSub.length) {
            sumArr.push(arrMain[i] + arrSub[i]);
        }
        // arrMain
        else {
            sumArr.push(arrMain[i]);
        }
    }
    console.log('before : ', sumArr);

    for (let i = 0; i < sumArr.length; i++) {
        if (sumArr[i] > 9) {
            quotient = Math.floor(sumArr[i] / 10);
            remainder = sumArr[i] % 10;

            sumArr[i] = remainder;
            // [10] -> [0, 1]
            if (typeof(sumArr[i+1]) === 'number') {
                sumArr[i+1] += quotient;
            }
            else {
                sumArr.push(quotient);
            }
        }
    }
    console.log('after  : ', sumArr);

    return sumArr
}

function makeNumArr(l) {
    let numArr = [];
    while(true) {
        numArr.push(l.val);
        if (!l.next) break;
        l = l.next;
    }
    return numArr;
}

var addTwoNumbers = function(l1, l2) {
    sumArr = arraySum(makeNumArr(l1), makeNumArr(l2));
    l = makeLinkedListFromArr(sumArr);
    return l;
};

printLinkedList(makeLinkedList(342));
printLinkedList(makeLinkedList(465));
result = addTwoNumbers(makeLinkedList(342), makeLinkedList(465));
printLinkedList(result);



// printLinkedList(makeLinkedList(10000000000000000001))
// printLinkedList(makeLinkedList(465))
// result = addTwoNumbers(makeLinkedList(1000000000000000000000000000001), makeLinkedList(465))
// printLinkedList(result)
